### PR TITLE
RUM-7102 feat: Add Time To Network Settled metric in RUM

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

📦⏱️ This PR introduces the **Time-to-Network-Settled (TTNS)** metric to measure the time from when a view becomes visible until all initial network resources have finished loading. This metric helps gauge the time it takes for a view to be fully displayed with all required resources.

The TTNS value is reported under these conditions:
- Resources starting within 100ms of the view becoming visible are considered "initial network resources."
- The TTNS value is available only after all initial resources load, and at least 100ms after the view starts.
- The value is not tracked if the view is stopped before resources finish loading.
- The value is tracked only if the app was in the "active" state during view loading.

**Example:** In following scenario, resources 1, 2, and 3 are "initial," and the completion of the last resource determines TTNS. Resource 4 is not "initial" as it started after the 100ms threshold.

Screenshot 2024-11-29 at 10 29 45

The TTNS value is reported in the `view.network_settled` attribute, in nanoseconds.

This is the counterpart to [Android PR #2392](https://github.com/DataDog/dd-sdk-android/pull/2392).

### How?

The metric is modeled as an object implementing the `TTNSMetricTracking` interface, which tracks key events related to TTNS (resource start, stop, and view stop) and provides a method to retrieve the metric value:

```swift
internal protocol TTNSMetricTracking {
    func trackResourceStart(at startDate: Date, resourceID: RUMUUID)
    func trackResourceEnd(at endDate: Date, resourceID: RUMUUID, resourceDuration: TimeInterval?)
    func trackResourceDropped(resourceID: RUMUUID)
    func trackViewWasStopped()
    func value(at time: Date, appStateHistory: AppStateHistory) -> TimeInterval?
}
```

The TTNS object is created within each `RUMViewScope` at the start of the view and passed to `RUMResourceScope` to track resources associated with that view.

This PR includes both unit and integration tests for TTNS, with integration tests located in `ViewLoadingMetricsTests`. These tests will also serve as a foundation for upcoming view loading metrics.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [x] Add CHANGELOG entry for user facing changes
- [ ] Add Objective-C interface for public APIs (see our [guidelines](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3157787243/RFC+-+Modular+Objective-C+Interface#Recommended-solution) [internal]) and run `make api-surface`)

### CODE REVIEW

**Review by mariedm — APPROVED** [review]
Well done! 💪 Just a couple of non-blocking questions and comments, but overall, it's nicely written, well-documented, and the test suite is robust.

File: DatadogRUM/Sources/DataModels/RUMDataModels.swift [hunk 1]
```diff
@@ -1827,6 +1830,7 @@ public struct RUMResourceEvent: RUMDataModel {
             case transferSize = "transfer_size"
             case type = "type"
             case url = "url"
+            case worker = "worker"
```
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogRUM/Sources/DataModels/RUMDataModels.swift:1833]: They are required as schema mixes multiple conventions (mainly snake_case and kebab-case), e.g.:
```swift
case analytics = "analytics"
// ...
case reactNative = "react-native"
// ...
case sessionSampleRate = "session_sample_rate"
```

File: DatadogRUM/Sources/RUMMetrics/TTNSMetric.swift [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    /// The `resourceDuration` is used if available; otherwise, the duration is calculated from the start to the end time.
+    ///
+    /// - Parameters:
+    ///   - endDate: The end time of the resource (device time, no NTP offset).
+    ///   - resourceID: The unique identifier for the resource.
+    ///   - resourceDuration: The resource duration, if available.
+    func trackResourceEnd(at endDate: Date, resourceID: RUMUUID, resourceDuration: TimeInterval?) {
+        guard isViewActive, let startDate = pendingResourcesStartDates[resourceID] else {
+            return // View was stopped or the resource was not tracked
+        }
```
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogRUM/Sources/RUMMetrics/TTNSMetric.swift:99]: Fair call, added ✅

File: DatadogRUM/Tests/RUMMetrics/TTNSMetricTests.swift [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+    func testWhenResourceEndsThenStarts() {
+        metric.trackResourceEnd(at: viewStartDate + 1, resourceID: .resource1, resourceDuration: nil)
+        metric.trackResourceStart(at: viewStartDate, resourceID: .resource1)
+        XCTAssertNil(
+            metric.value(at: .distantFuture, appStateHistory: .mockAppInForeground(since: .distantPast)),
+            "Metric value should be nil if resource ends before it starts."
+        )
+    }
+
```
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogRUM/Tests/RUMMetrics/TTNSMetricTests.swift:348]: 👍 Yes! Added ✅

File: DatadogRUM/Sources/RUMMetrics/TTNSMetric.swift [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    /// Tracks the completion of a resource identified by its `resourceID`.
+    /// The `resourceDuration` is used if available; otherwise, the duration is calculated from the start to the end time.
+    ///
+    /// - Parameters:
+    ///   - endDate: The end time of the resource (device time, no NTP offset).
+    ///   - resourceID: The unique identifier for the resource.
+    ///   - resourceDuration: The resource duration, if available.
+    func trackResourceEnd(at endDate: Date, resourceID: RUMUUID, resourceDuration: TimeInterval?) {
```
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogRUM/Sources/RUMMetrics/TTNSMetric.swift:96]: The "start" and "end" times for RUM resource are approximations based on network request timings derived from the `resume()` and `completion {}` in URLSession.

For a more accurate representation, RUM listens to [`URLSessionTaskMetrics`](https://developer.apple.com/documentation/foundation/urlsessiontaskmetrics) delivered to the session delegate. In addition to [various temporal metrics](https://developer.apple.com/documentation/foundation/urlsessiontasktransactionmetrics#3162615) such as DNS, TCP, and TLS timings, `URLSessionTaskMetrics` also provides the fetch duration. This fetch duration serves as a more precise alternative for the RUM resource "duration," which is now tracked here for the TTNS metric.

File: DatadogRUM/Tests/RUMMonitor/Scopes/RUMViewScopeTests.swift [hunk 5]
```diff
@@ -2595,7 +2595,7 @@ class RUMViewScopeTests: XCTestCase {
         XCTAssertEqual(event.view.resource.count, 1, "After dropping 1 Resource event (out of 2), View should record 1 Resource")
         XCTAssertEqual(event.view.action.count, 0, "After dropping a User Action event, View should record no actions")
         XCTAssertEqual(event.view.error.count, 0, "After dropping an Error event, View should record 0 Errors")
-        XCTAssertEqual(event.dd.documentVersion, 3, "After starting the application, stopping the view, starting/stopping one resource out of 2, discarding a user action and an error, the View scope should have sent 3 View events.")
+        XCTAssertEqual(event.dd.documentVersion, 4, "It should create 4 view update.")
```
Comment (ncreated) [inline DatadogRUM/Tests/RUMMonitor/Scopes/RUMViewScopeTests.swift:2598]: 👍 This is expected as now we also cause view updates when events are dropped. This doesn't change what is sent to RUM intake as redundant updates are filtered before upload.

**Comment by ncreated** [comment]
@mariedm Thanks for review! All feedback addressed 👌.

I'm requesting another approval because there was small addition of new logic in the last commit (https://github.com/DataDog/dd-sdk-ios/pull/2125/commits/fd4e9165274b5bd5dfa9813f287a6d4bdc8e65fc) following recent spec changes. We now only count resources that are sent to RUM, excluding any resources that are dropped by the user with event mapper API.

File: DatadogRUM/Sources/RUMMetrics/TTNSMetric.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (mariedm) [reply DatadogRUM/Sources/RUMMetrics/TTNSMetric.swift:96]: Got it, thanks for the explanation!

**Review by mariedm — APPROVED** [review]
LGTM! 🚀

**Pull request merged**